Script improved to get closer to production use

aalaesar opened the pull request:
The script has been improved in various ways to get closer to production requirements  but is still retro-compatible for previous use.
### Major changes:

- LibreOffice major/stable version can be specified or detected in repository
- Build options for Loolwsd
  - enabled-debug added as anoption
  - changed default log file to /var/log/loolwsd.log
- Added **idempotence** to most task.
  - the script is able to be run several times on the same host and only alter what can be changed.
- Changed how the dialog behave:
  - Only appears if LibreOffice has to be compiled
  - Show an estimated compilation time depending on number of core available
- Changed Installation & test step:
  - only run after a Loolwsd successfull compilation
  - test step do not requires sudo privileges anymore.
- Changed log generation:
  - log file has a Date (YYYYMMDD-HHMM) Prefix
  - All the script's output but the dialog prompt is now logged, stderr included
### Misc
- more parameters:
  - `lo_src_repo` & `lo_version` for choosing a libreoffice version to use
  - **added** `sh_interactive `variable to toggle dialog (for scripting use)
  - `lo_forcebuild` & `lool_forcebuild` to force compilation.
- Renamed somes variable for easy-to-recognize names
- "Structural" comments && average compilation time for lo, poco & loolwsd

husisusi commented:
Wow! you rewritten almost everything.
Thank you for your work!
But it dose not work:
[checking for cups/cups.h... no
configure: error: Could not find CUPS. Install libcups2-dev or cups-devel.
Error running configure at ./autogen.sh line 281.
make: *** No targets specified and no makefile found.  Stop.]

husisusi commented:
Can you Pleas check?

aalaesar commented:
Well I run it several times with success on debian jessie and one more time and  just after making the pull request for good measure.
I suppose you tested it on Ubuntu ? I'll make a run on it.
How is configured your test env ?

By the way, I can still push commit  to my branch and therefore to the PR. I'm not sure closing it is required for fixing it. ;)

husisusi commented:
Yes, I tested it On Ubuntu 16.04 KVM VM.
It should working on Debian and Ubuntu...
Thanks again!

aalaesar commented:
Ok I also thought this way. That mostly why I kept testing on debian. Lazy me 😅...
I didn't remove apt packages in the script anyway :/.

aalaesar commented:
Configure error fixed.
added exit condition for configure & make steps
rebased all the branch to resolve conflicts

I realised my designs in choosing the version to download is bad.
I'll do one more commit before you can merge. but you still can test it if you want.

aalaesar commented:
done the fix
I leave it to you 😉

aalaesar commented:
Just worked on the logs mecanism. that will be my final commit for this PR.

husisusi commented:
I just tested it on Ubuntu 16.04.
loolwsd service is running and port 9980 is up. But if I try to access https://localhost:9980/loleaflet.....aflet.html?file_path=file:///opt/online/test/data/hello-world.odt I get:
"Service is unavailable. Please try again later and report to your administrator if the issue persists."

aalaesar commented:
ho yes this one,
this is a bad message for a "file not found" error
I met it but  I didn't had time to investigate further:
[i did now, i was my fault]

aalaesar commented:
I found what happen:
hello-world.odt is a copy of hello.odt
the copy is done by the `run all` target ...
```make
run: all ${abs_top_builddir}/jails ${abs_top_builddir}/systemplate/system_stamp
        @echo "Launching loolwsd"
        @cp $(abs_top_srcdir)/test/data/hello.odt $(abs_top_srcdir)/test/data/hello-world.odt
        @echo
        ./loolwsd --o:sys_template_path="${abs_top_builddir}/systemplate" --o:lo_template_path="/opt/libreoffice/instdir" \
                          --o:child_root_path="${abs_top_builddir}/jails" --o:storage.filesystem[@allow]=true \
                          --o:admin_console.username=admin --o:admin_console.password=admin
```
.. which I removed from the script because it requires sudo privileges
i'll just add back the copy itself into the script

husisusi commented:
Thanks, Testing....

aalaesar commented:
Idempotence is very useful for testing small fixes like this one.
If you test on the same host you previously used,the script will skip LO compilation.
but because loolwsd is already compiled while installation & test parts are run only after compiling loolwsd (and the last fix is here), you need to set
`lool_forcebuild=true` in the script to force loowsd compilation.
The test would take just minutes instead of hours.
=> I keep a snapshot with LO compiled to test faster.

husisusi commented:
Can I add :
Written by: Subhi H. & Marc C. ?

aalaesar commented:
Yes
Thanks ! 👍
